Address some compiler warnings

abartlet opened the pull request:
These compiler warnings come via Samba.

Some were already submitted in #354 but I re-raise them with the justifications requested as far as I've been able to reproduce on Ubuntu 16.04 gcc 5.4.0

I realise not all the patches and chunks have a justification, but increasing the type width in (eg) asn1.c is pretty low-risk. 

It would be very helpful if these could be accepted upstream as it will reduce the number of patches we need to maintain in our lorikeet-heimdal fork.

Thanks!

jaltman commented on lib/krb5/krbhst.c:
```diff
@@ -353,13 +353,13 @@ krb5_krbhst_format_string(krb5_context context, const krb5_krbhst_info *host,
-	snprintf(portstr, sizeof(portstr), ":%d", host->port);
+	snprintf(portstr, sizeof(portstr), ":%hd", host->port);
     snprintf(hostname, hostlen, "%s%s%s", proto, host->hostname, portstr);
     return 0;
```
I would prefer to see this rewritten to remove portstr[] entirely

  if (host->port != host->def_port)
    snprintf(hostname, hostlen, "%s%s%hd", proto, host->hostname, host->port);
  else
    snprintf(hostname, hostlen, "%s%s", proto, host->hostname);

lhoward replied:
%hd is widely supported I presume

lhoward replied:
I applied Jeff's fix

nicowilliams replied:
That works.  `%hd` is still not needed (since it's an int), but it's fine.

lhoward replied:
I couldn't push to @abartlet's branch to update this PR, but the fix I committed to master uses %d and an int cast.

jaltman commented on lib/gssapi/mech/gss_set_cred_option.c:
```diff
@@ -41,7 +41,7 @@ gss_set_cred_option (OM_uint32 *minor_status,
 	OM_uint32	major_status = GSS_S_COMPLETE;
 	struct _gss_mechanism_cred *mc;
-	int one_ok = 0;
+	OM_uint32 one_ok = 0;
```
one_ok is not a protocol variable.  Its a local variable used as a boolean flag.  Its proper type is 'int'.

abartlet replied:
Thanks, I agree.  I can't see how this was ever an issue, I'll drop that hunk.

abartlet replied:
Looping back to this one, there is no krb5_boolean available in gsspai, is Heimdal willing to use C99 bool yet?

The issue is that on Ubuntu 14.04 with strict compiler options and -O3 we get:


==> /builds/catalyst-samba/samba/samba-o3.stderr <==
../source4/heimdal/lib/gssapi/mech/gss_set_cred_option.c: In function ‘gss_set_cred_option’:
../source4/heimdal/lib/gssapi/mech/gss_set_cred_option.c:112:5: error: assuming signed overflow does not occur when simplifying conditional to constant [-Werror=strict-overflow]
  if (one_ok) {
     ^
cc1: some warnings being treated as errors

If I put that back into the commit message, can I have this hunk accepted?

jaltman replied:
This warning has to be a compiler bug.    
I'm curious, what happens if you change that to 

   if (!one_ok)
       return major_status;

   *minor_status = 0;
   return GSS_S_COMPLETE;

abartlet commented:
I've addressed the feedback.

Is this OK to merge now?

Thanks,

Andrew Bartlett

jaltman commented:
The Travis builders hung when running the tests.

jaltman commented on lib/krb5/krbhst.c:
```diff
@@ -353,14 +353,15 @@ krb5_krbhst_format_string(krb5_context context, const krb5_krbhst_info *host,
-    if(host->port != host->def_port)
-	snprintf(portstr, sizeof(portstr), ":%d", host->port);
-    snprintf(hostname, hostlen, "%s%s%s", proto, host->hostname, portstr);
+    if(host->port != host->def_port) {
```
Although there are lots of instances of `if()` in the tree instead of `if ()` our preferred style is a space after `if` as `if` is not a function.   We are not going to globally fix this because we do not want to destroy the usefulness of git blame.

abartlet replied:
Thanks, is there a document on this somewhere?  

(This is Samba's style also, for what it is worth)

abartlet commented:
> The Travis builders hung when running the tests.

I've isolated the patch that does this, and am very glad such tests exist.  I've cut Andreas's patch down to the parts I could reproduce on Ubuntu 16.04 and which I included in the commit message and re-pushed.

abartlet commented:
Sorry for the drama, I finally worked out what was going on:  I botched the first patch that I was asked to re-write, rather than cherry-pick.  

To be clear, it wasn't Andreas's patch, but I've left it cut-down to the parts I could reproduce the failure for.

jaltman commented on lib/krb5/keytab_any.c:
```diff
@@ -222,7 +222,7 @@ any_remove_entry(krb5_context context,
     struct any_data *a = id->data;
     krb5_error_code ret;
-    int found = 0;
+    size_t found = 0;
```
'found' should be acceptable as an 'int'.   Its used as a boolean either zero or non-zero.

Given its usage setting `found = 1;` instead of `found++;` at line 229 is probably the correct fix.

jaltman commented on lib/krb5/config_file.c:
```diff
@@ -1081,7 +1081,7 @@ krb5_config_vget_strings(krb5_context context,
 {
     char **strings = NULL;
-    int nstr = 0;
+    size_t nstr = 0;
```
This is fine.

abartlet commented:
Of course it is a 6 year old compiler, I'm just trying not to carry
local patches.

We can go around and around forever, but I'm just wasting your time and
not getting any closer to reducing the patches I carry locally.

What do you want to do, for the global picture, because this back and
forth is not productive for either of us. 

I don't get to choose the compiler Samba's autobuild runs with, and for
now it is this one, for better or worse.

Neither can I drop the patches, much as would like, because it then I
can't update Samba's copy.  I have far, far bigger issues to address
but we are both bogged down in this, looking for 'perfect'.

So which do you want:
 - the patches in our branch, so we never get to using upstream master
or
 - change the type of the boolean to uint32_t
or
 - change the type of the boolean to bool?

Thanks,

Andrew Bartlett

abartlet commented:
To try and move this further along, I'll leave the gss_set_cred_option.c change in our branch until we can get Samba off the old compiler for our CI.  Updated patch set pushed, which should, I hope, have addressed all the review comment provided.

lhoward commented:
Merged with stylistic fixes

abartlet commented:
man 3 snprintf suggests at least since SUSv2:

SUSv2 specified only the length modifiers h (in  hd,
       hi, ho, hx, hX, hn) and l (in ld, li, lo, lx, lX, ln, lc, ls)
and L (in Le, LE, Lf, Lg, LG).

abartlet commented:
> Merged with stylistic fixes

Thanks!

lhoward commented:
@nicowilliams suggested int cast instead of %hd so went with that.
